File: benchmarks/src/com/sun/labs/kanban/richards_deutsch_no_acc/c++/richards.cc
/* richards.cc - Richards Benchmark in C++ */
/* Urs Hoelzle 2/2/89 */
/* mods, Mario Wolczko 4/2/96 */
/* mods, Alex Jacoby, 8/9/96 */

#include <time.h>
#include <stdio.h>
#include <ctype.h>
#include <assert.h>
#include <stdlib.h>
#include <sys/types.h>

// Some compilers predefine this; if so, comment it out.
// typedef enum { false, true } bool;

typedef enum {DevicePacket, WorkPacket} PacketKind;

typedef enum {Idler, Worker, HandlerA, HandlerB, DeviceA, DeviceB} Identity;

#define NTASKS 6                        /* # elements of Identity */
				 	/* = # of tasks		  */
#define NoWork NULL
#define NoTask NULL

/*
// Packet
*/

class Packet {

public:
  Packet *link;		// next packet in queue
  Identity ident;	// Idle, Worker, DeviceA, ...
  PacketKind kind;	// DevicePacket or WorkPacket
  int datum;
  char data[4];

public:
  Packet(Packet *l, Identity id, PacketKind k);

  Packet *AddToList(Packet *list);
};

Packet::Packet(Packet *l, Identity id, PacketKind k)
{
  link = l;
  ident = id;
  kind = k;
  datum = 1;
  for(int i = 0; i < 4; i++)
    data[i] = 0;
}

/*
// AddToList - list utility (append this at end of list, return head)
*/
Packet *Packet::AddToList(Packet *list)
{
  link = NoWork;
  if (list == NoWork)
    list = this;
  else {
    Packet *p, *next;
    p = list;
    while (next = p->link, next != NoWork)
      p = next;
    p->link = this;
  }    
  return list;
}



/*
// DeviceTaskRec
*/

class DeviceTaskRec {
public:
  Packet *pending;
public:
  DeviceTaskRec() { pending = NoWork; }
};



/*
// IdleTaskRec 
*/

class IdleTaskRec {
public:
  int control, count;
public:
  IdleTaskRec() { control = 1; count = 10000; }
};


/*
// HandlerTaskRec
*/

class HandlerTaskRec {
public:
  Packet *workIn, *deviceIn;
public:
  HandlerTaskRec() { workIn = deviceIn = NoWork; }

  Packet *  WorkInAdd(Packet *p) {
    return workIn = p->AddToList(workIn); }
  Packet *DeviceInAdd(Packet *p) {
    return deviceIn = p->AddToList(deviceIn); }
};


/*
// WorkerTaskRec
*/

class WorkerTaskRec {
public:
  Identity destination;
  int count;
public:
  WorkerTaskRec() { destination = HandlerA; count = 0; }
};



/*
// TaskState
*/

class TaskState {
public:
  bool packetPending, taskWaiting, taskHolding;
	
public:
  /* initializing */
  TaskState() {
    packetPending = true;
    taskWaiting = taskHolding = false;
  }
  void PacketPending() {
    packetPending = true;
    taskWaiting = taskHolding = false;
  }
  void Waiting() {
    packetPending = taskHolding = false;
    taskWaiting = true;
  }
  void Running() {
    packetPending = taskWaiting = taskHolding = false;
  }
  void WaitingWithPacket() {
    packetPending = taskWaiting = true; taskHolding = false;
  }

  /* testing */ 
  bool IsTaskHoldingOrWaiting() {
    return bool(taskHolding || !packetPending && taskWaiting);
  }
  bool IsWaitingWithPacket() {
    return bool(packetPending && taskWaiting && !taskHolding);
  }
};

/*
// TaskControlBlock
*/

class TaskControlBlock : public TaskState {

public:
  TaskControlBlock *link;
  Identity ident;
  int priority;
  Packet *input;
  void *handle;
  
public:
  /* initializing */
  TaskControlBlock(TaskControlBlock *l, Identity id, int prio,
		   Packet *initialWork, TaskState *initialState,
		   void *privateData);

  /* scheduling */
  TaskControlBlock *AddPacket(Packet *p, TaskControlBlock *old);
  virtual TaskControlBlock *ActionFunc(Packet *p, void *handle);
  TaskControlBlock *RunTask();
};

/*
// DeviceTCB 
*/

class DeviceTCB : public TaskControlBlock {
public:
  DeviceTCB(TaskControlBlock *l, Identity id, int prio,
	    Packet *initialWork, TaskState *initialState,
	    void *privateData) :
    TaskControlBlock(l, id, prio, initialWork, initialState, privateData) {}

  TaskControlBlock *ActionFunc(Packet *p, void *handle);
};


/*
// HandlerTCB
*/

class HandlerTCB : public TaskControlBlock {
public:
  HandlerTCB(TaskControlBlock *l, Identity id, int prio,
	     Packet *initialWork, TaskState *initialState,
	     void *privateData) :
    TaskControlBlock(l, id, prio, initialWork, initialState, privateData) {}

  TaskControlBlock *ActionFunc(Packet *p, void *handle);
};


/*
// IdlerTCB
*/

class IdlerTCB : public TaskControlBlock {
public:
  IdlerTCB(TaskControlBlock *l, Identity id, int prio,
	   Packet *initialWork, TaskState *initialState,
	   void *privateData) :
    TaskControlBlock(l, id, prio, initialWork, initialState, privateData) {}

  TaskControlBlock *ActionFunc(Packet *p, void *handle);
};


/*
// WorkerTCB
*/

class WorkerTCB : public TaskControlBlock {
public:
  WorkerTCB(TaskControlBlock *l, Identity id, int prio,
	    Packet *initialWork, TaskState *initialState,
	    void *privateData) :
    TaskControlBlock(l, id, prio, initialWork, initialState, privateData) {}

  TaskControlBlock *ActionFunc(Packet *p, void *handle);
};


/*
// RBench class definition
*/

class RBench {
public:
  TaskControlBlock *taskList;
  TaskControlBlock *currentTask;
  Identity          currentTaskIdent;
  TaskControlBlock *taskTable[NTASKS];
  int layout;
  int holdCount;
  int queuePacketCount;

  /* creation */
  void CreateDevice (Identity id, int prio,  Packet *work, TaskState *state);
  void CreateHandler(Identity id, int prio,  Packet *work, TaskState *state);
  void CreateIdler  (Identity id, int prio,  Packet *work, TaskState *state);
  void CreateWorker (Identity id, int prio,  Packet *work, TaskState *state);
  void EnterTask    (Identity id, TaskControlBlock *t);

  /* scheduling */
  void Schedule();
  
  /* initializing */
  void InitScheduler();
  void InitTrace();
public:
  /* task management */
  TaskControlBlock *FindTask(Identity id);
  TaskControlBlock *HoldSelf();
  TaskControlBlock *QueuePacket(Packet *p);
  TaskControlBlock *Release(Identity id);
  TaskControlBlock *Wait();

  /* tracing */
  bool tracing;
  void Trace(Identity id);
  
  void Start(bool askTrace);
};



class RBench bm;		// benchmark currently executing


/*
// TaskControlBlock
*/

TaskControlBlock::TaskControlBlock(TaskControlBlock *l,
				   Identity id,
				   int prio,
				   Packet *initialWork,
				   TaskState *initialState,
				   void *privateData)
{
  link = l;
  ident = id;
  priority = prio;
  input = initialWork;
  packetPending = initialState->packetPending;
  taskWaiting = initialState->taskWaiting;
  taskHolding = initialState->taskHolding;
  handle = privateData;
}


TaskControlBlock *TaskControlBlock::AddPacket(Packet *p,
                                              TaskControlBlock *oldTask)
{
  if (input == NoWork) {
    input = p;
    packetPending = true;
    if (priority > oldTask->priority)
      return this;
  } else {
    p->AddToList(input);
  }
  return oldTask;
}


TaskControlBlock *TaskControlBlock::RunTask()
{
  Packet *msg;

  if (IsWaitingWithPacket()) {
    msg = input;
    input = input->link;
    if (input == NoWork)
      Running();
    else
      PacketPending();
  } else {
    msg = NoWork;
  }
  return ActionFunc(msg, handle);
}


/*
// action functions
*/

TaskControlBlock *TaskControlBlock::ActionFunc(Packet *, void *)
{
  printf("***error: virtual TaskControlBlock.ActionFunc called!\n");
  return NoTask;
}


TaskControlBlock *DeviceTCB::ActionFunc(Packet *p, void *h)
{
  DeviceTaskRec *data = (DeviceTaskRec *)h;

  if (p == NoWork) {
    p = data->pending;
    if (p == NoWork)
      return bm.Wait();
    else {
      data->pending = NoWork;
      return bm.QueuePacket(p);
    }
  } else {
    data->pending = p;
    if (bm.tracing)
      bm.Trace((Identity)p->datum);
    return bm.HoldSelf();
  }
}


TaskControlBlock *HandlerTCB::ActionFunc(Packet *p, void *h)
{
  HandlerTaskRec *data = (HandlerTaskRec *)h;
  Packet *work, *devicePacket;
  int count;

  if (p != NoWork) {
    if (p->kind == WorkPacket)
      data->WorkInAdd(p);
    else
      data->DeviceInAdd(p);
  }
  work = data->workIn;
  if (work == NoWork)
    return bm.Wait();

  count = work->datum;
  if (count > 4) {
    data->workIn = work->link;
    return bm.QueuePacket(work);
  } else {
    devicePacket = data->deviceIn;
    if (devicePacket == NoWork)
      return bm.Wait();
    else {
      data->deviceIn = devicePacket->link;
      devicePacket->datum = work->data[count-1];
      work->datum = count + 1;
      return bm.QueuePacket(devicePacket);
    }
  }
}


TaskControlBlock *IdlerTCB::ActionFunc(Packet *, void *h)
{
  IdleTaskRec *data = (IdleTaskRec *)h;

  if (--data->count == 0)
    return bm.HoldSelf();
  else if ((data->control & 1) == 0) {
    data->control /= 2;
    return bm.Release(DeviceA);
  } else {
    data->control = (data->control / 2) ^ 53256;
    return bm.Release(DeviceB);
  }
}


TaskControlBlock *WorkerTCB::ActionFunc(Packet *p, void *h)
{
  WorkerTaskRec *data = (WorkerTaskRec *)h;

  if (p == NoWork)
    return bm.Wait();
  
  Identity dest = (data->destination == HandlerA) ? HandlerB : HandlerA;
  data->destination = dest;
  p->ident = dest;
  p->datum = 1;
  for(int i = 0; i < 4; i++) {
    if (++data->count > 26) data->count = 1;
    p->data[i] = 'A' + data->count;
  }
  return bm.QueuePacket(p);
}




/*
// creation
*/

void RBench::CreateDevice (Identity id, int prio,  Packet *work,
                           TaskState *state)
{   DeviceTaskRec *data;
    TaskControlBlock *t;

    data = new DeviceTaskRec;
    t = new DeviceTCB(taskList, id, prio, work, state, data);
    EnterTask(id, t);
}


void RBench::CreateHandler(Identity id, int prio,  Packet *work,
                           TaskState *state)
{   HandlerTaskRec *data;
    TaskControlBlock *t;

    data = new HandlerTaskRec;
    t = new HandlerTCB(taskList, id, prio, work, state, data);
    EnterTask(id, t);
}


void RBench::CreateIdler  (Identity id, int prio,  Packet *work,
                           TaskState *state)
{   IdleTaskRec *data;
    TaskControlBlock *t;

    data = new IdleTaskRec;
    t = new IdlerTCB(taskList, id, prio, work, state, data);
    EnterTask(id, t);
}


void RBench::CreateWorker (Identity id, int prio,  Packet *work,
                           TaskState *state)
{   WorkerTaskRec *data;
    TaskControlBlock *t;

    data = new WorkerTaskRec;
    t = new WorkerTCB(taskList, id, prio, work, state, data);
    EnterTask(id, t);
}


void RBench::EnterTask(Identity id, TaskControlBlock *t)
{
    taskList = t;
    taskTable[id] = t;
}


/*
//private
*/

TaskControlBlock *RBench::FindTask(Identity id)
{   TaskControlBlock *t;

    t = taskTable[id];
    if (t == NULL) printf("***error: FindTask failed! ");
    return t;
}


TaskControlBlock *RBench::HoldSelf()
{
    holdCount++;
    currentTask->taskHolding = true;
    return currentTask->link;
}


TaskControlBlock *RBench::QueuePacket(Packet *p)
{   TaskControlBlock *t;

    t = FindTask(p->ident);
    queuePacketCount++;
    p->link = NoWork;
    p->ident = currentTaskIdent;
    return t->AddPacket(p, currentTask);
}


TaskControlBlock *RBench::Release(Identity id)
{   TaskControlBlock *t;

    t = FindTask(id);
    t->taskHolding = false;
    return t->priority > currentTask->priority ? t : currentTask;
}


void RBench::Trace(Identity id)
{
    if (--layout != 0) { printf("\n"); layout = 50; } 
    printf("%d", id + 1);
}


TaskControlBlock *RBench::Wait()
{
    currentTask->taskWaiting = true;
    return currentTask;
}


/*
//scheduling
*/

void RBench::Schedule()
{
    currentTask = taskList;
    while (currentTask != NoTask) {
         if (currentTask->IsTaskHoldingOrWaiting()) 
              currentTask = currentTask->link;
         else {
              currentTaskIdent = currentTask->ident;
              if (tracing) Trace(currentTaskIdent);
              currentTask = currentTask->RunTask();
         }
    }
}

/*
//initializing
*/

void RBench::InitScheduler()
{
    queuePacketCount = 0;
    holdCount = 0;
    for (int i = 0; i < NTASKS; i++)
      taskTable[i] = NoTask;
    taskList = NoTask;
}


void RBench::InitTrace()
{   char c;

    printf("Trace (y/n)? ");
    c = getchar();
    tracing = bool(_toupper(c) == 'Y');
}


void RBench::Start(bool trace) {
    TaskState *t;
    Packet *workQ;

    if (trace) InitTrace(); else tracing = false;
    InitScheduler();
    t = new TaskState; t->Running();				// Idler
    CreateIdler(Idler, 0, NoWork, t);

    workQ = new Packet(NoWork, Worker, WorkPacket);		// Worker
    workQ = new Packet(workQ , Worker, WorkPacket);
    t = new TaskState; t->WaitingWithPacket();
    CreateWorker(Worker, 1000, workQ, t);

    workQ = new Packet(NoWork, DeviceA, DevicePacket);		// HandlerA
    workQ = new Packet(workQ , DeviceA, DevicePacket);
    workQ = new Packet(workQ , DeviceA, DevicePacket);
    t = new TaskState; t->WaitingWithPacket();
    CreateHandler(HandlerA, 2000, workQ, t);

    workQ = new Packet(NoWork, DeviceB, DevicePacket);		// HandlerB
    workQ = new Packet(workQ , DeviceB, DevicePacket);
    workQ = new Packet(workQ , DeviceB, DevicePacket);
    t = new TaskState; t->WaitingWithPacket();
    CreateHandler(HandlerB, 3000, workQ, t);

    t = new TaskState; t->Waiting();				// DeviceA
    CreateDevice(DeviceA, 4000, NoWork, t);
    t = new TaskState; t->Waiting();				// DeviceB
    CreateDevice(DeviceB, 5000, NoWork, t);

    Schedule();

    if (queuePacketCount == 23246 && holdCount == 9297)
      ; // correct
    else {
      printf("error: richards results are incorrect\n");
      exit(1);
    }
}

#define ITER  100		/* # of iterations in main loop  */

int main()
{
  printf("Richards #3 starting\n");
  clock_t start_time= clock();
  for (int i = 0; i < ITER; i++)
    bm.Start(false);
  clock_t end_time = clock();
  double total_s= (end_time - start_time) / double(CLOCKS_PER_SEC);
  double avg_ms= total_s / ITER * 1000;
  printf("elapsed time for %d iterations: %.3g seconds\n", ITER, total_s);
  printf("average time per iteration: %.3g ms\n", avg_ms);
}

File: benchmarks/src/com/sun/labs/kanban/richards_deutsch_no_acc/Richards.java
//  Based on original version written in BCPL by Dr Martin Richards
//  in 1981 at Cambridge University Computer Laboratory, England
//  and a C++ version derived from a Smalltalk version written by
//  L Peter Deutsch.
//  Java version:  Copyright (C) 1995 Sun Microsystems, Inc.
//  Translation from C++, Mario Wolczko
//  Outer loop added by Alex Jacoby
//  Modified to run on Squawk platform by Erica Glynn (03/02/2005).

package com.sun.labs.kanban.richards_deutsch_no_acc;

//----- Packet -------------------------------------------------------

class Packet {
  static final int BUFSIZE = 4;

  Packet link;
  int id;
  int kind;
  int datum;
  int[] data = new int[BUFSIZE];

  Packet(Packet l, int i, int k) {
    link = l;
    id = i;
    kind = k;
    datum = 0;
    for (int j = 0;  j < BUFSIZE;  j++) 
      data[j] = 0;
  }

  Packet append_to(Packet list) {
    link = null;
    if (list == null) 
      return this;
    else {
      Packet p = list;
      Packet next = p.link;
      while (next != null) {
        p = next;
	next = p.link;
      }
      p.link = this;
      return list;
    }
  }

}

//----- Task Records------------------------------

abstract class TaskRec { } // so we have a common type for all task records

class DeviceTaskRec extends TaskRec {
  Packet pending;

  DeviceTaskRec()           { pending = null; }
}


class IdleTaskRec extends TaskRec {
  int control, count;

  IdleTaskRec() { control = 1; count = 10000; }
}


class HandlerTaskRec extends TaskRec {
  Packet workIn, deviceIn;

  HandlerTaskRec() { workIn = deviceIn = null; }

  Packet   WorkInAdd(Packet p) { return workIn = p.append_to(workIn); }
  Packet DeviceInAdd(Packet p) { return deviceIn = p.append_to(deviceIn); }
}



class WorkerTaskRec extends TaskRec {
  int destination;
  int count;

  WorkerTaskRec() { destination = Richards.I_HANDLERA; count = 0; }
}


//----- Task ---------------------------------------------------------

class TaskState {

  boolean packetPending, taskWaiting, taskHolding;

  TaskState() {
    packetPending = true;
    taskWaiting = false;
    taskHolding = false;
  }

  TaskState PacketPending() {
    packetPending = true;
    taskWaiting = taskHolding = false;
    return this;
  }
  TaskState Waiting() {
    packetPending = taskHolding = false;
    taskWaiting = true;
    return this;
  }
  TaskState Running() {
    packetPending = taskWaiting = taskHolding = false;
    return this;
  }
  TaskState WaitingWithPacket() {
    packetPending = taskWaiting = true; taskHolding = false;
    return this;
  }

  /* testing */ 
  boolean IsTaskHoldingOrWaiting() {
    return taskHolding || !packetPending && taskWaiting;
  }
  boolean IsWaitingWithPacket() {
    return packetPending && taskWaiting && !taskHolding;
  }
}

abstract class Task extends TaskState {

  static int layout = 0;

  Task link;
  int id;
  int pri;
  Packet wkq;
  TaskRec handle;

  static final int TaskTabSize = 10;
  static Task[] taskTab = new Task[TaskTabSize];
  static Task taskList;

  static final boolean tracing = false;
  static int holdCount = 0;
  static int qpktCount = 0;

  Task(int i, int p, Packet w, TaskState initialState, TaskRec r) {
    link = taskList;
    id = i;
    pri = p;
    wkq = w;
    packetPending = initialState.packetPending;
    taskWaiting = initialState.taskWaiting;
    taskHolding = initialState.taskHolding;
    handle = r;
    taskList = this;
    taskTab[i] = this;
  }

  abstract Task fn(Packet pkt, TaskRec r);

  private Task AddPacket(Packet p, Task old) {
    if (wkq == null) {
      wkq = p;
      packetPending = true;
      if (pri > old.pri)
        return this;
    } else {
      p.append_to(wkq);
    }
    return old;
  }

  Task RunTask() {
    Packet msg;

    if (IsWaitingWithPacket()) {
      msg = wkq;
      wkq = wkq.link;
      if (wkq == null)
	Running();
      else
	PacketPending();
    } else {
      msg = null;
    }
    return fn(msg, handle);
  }

  protected Task waitTask() {
    taskWaiting = true;
    return this;
  }

  protected Task hold() {
    ++holdCount;
    taskHolding = true;
    return link;
  }

  protected Task release(int i) {
    Task t = findtcb(i);
    t.taskHolding = false;
    return t.pri > pri ? t : this;
  }

  protected Task qpkt(Packet pkt) {
    Task t = findtcb(pkt.id);
    qpktCount++;
    pkt.link = null;
    pkt.id = id;
    return t.AddPacket(pkt, this);
  }

  static Task findtcb(int id) {
    Task t = Task.taskTab[id];
    if (t == null) 
      System.out.println("\nBad task id " + id);
    return t;
  }

  static void trace(char a) {
    if (--layout <= 0) {
      System.out.println();
      layout = 50;
    }
    System.out.print(a);
  }

}

//----- DeviceTask ---------------------------------------------------

class DeviceTask extends Task
{
  DeviceTask(int i, int p, Packet w, TaskState s, TaskRec r) {
    super(i, p, w, s, r);
  }

  Task fn(Packet pkt, TaskRec r) {
    DeviceTaskRec d = (DeviceTaskRec)r;
    if (pkt == null) {
      pkt = d.pending;
      if (pkt == null) 
        return waitTask();
      else {
	d.pending = null;
	return qpkt(pkt);
      }
    } else {
      d.pending = pkt;
      if (tracing) trace((char)pkt.datum);
      return hold();
    }
  }
}


//----- HandlerTask --------------------------------------------------

class HandlerTask extends Task
{
  HandlerTask(int i, int p, Packet w, TaskState s, TaskRec r) {
    super(i, p, w, s, r);
  }

  Task fn(Packet pkt, TaskRec r) {
    HandlerTaskRec h = (HandlerTaskRec)r;
    if (pkt != null) {
      if (pkt.kind == Richards.K_WORK)
	h.WorkInAdd(pkt);
      else
        h.DeviceInAdd(pkt);
    }
    Packet work = h.workIn;
    if (work == null)
      return waitTask();

    int count = work.datum;

    if (count >= Packet.BUFSIZE) {
      h.workIn = work.link;
      return qpkt(work);
    }

    Packet dev = h.deviceIn;
    if (dev == null)
      return waitTask();

    h.deviceIn = dev.link;
    dev.datum = work.data[count];
    work.datum = count + 1;
    return qpkt(dev);
  }
}


//----- IdleTask -----------------------------------------------------

class IdleTask extends Task 
{
  IdleTask(int i, int a1, int a2, TaskState s, TaskRec r) {
    super(i, 0, null, s, r);
  }

  Task fn(Packet pkt, TaskRec r) {
    IdleTaskRec i = (IdleTaskRec)r;

    if (--i.count == 0) {
      return hold();
    } else if ((i.control & 1) == 0) {
      i.control /= 2;
      return release(Richards.I_DEVA);
    } else {
      i.control = (i.control / 2) ^ 0XD008;
      return release(Richards.I_DEVB);
    }
  }

}


//----- WorkTask -----------------------------------------------------

class WorkTask extends Task 
{
  WorkTask(int i, int p, Packet w, TaskState s, TaskRec r) {
    super(i, p, w, s, r);
  }

  Task fn(Packet pkt, TaskRec r) {
    WorkerTaskRec w = (WorkerTaskRec)r;

    if (pkt == null)
      return waitTask();

    int dest = (w.destination == Richards.I_HANDLERA
		? Richards.I_HANDLERB
		: Richards.I_HANDLERA);
    w.destination = dest;
    pkt.id = dest;
    pkt.datum = 0;
    for (int i = 0; i < Packet.BUFSIZE; i++) { 
      if (++w.count > 26) w.count = 1;
      pkt.data[i] = 'A' + w.count - 1;
    }
    return qpkt(pkt);
  }
}


//----- Richards -----------------------------------------------------


public class Richards
{
  private long total_ms;
  public long getRunTime() { return total_ms; }

  public static void main(String[] args) {
    (new Richards()).inst_main(args);
  }

  static int iterations = 10;

  public void inst_main(String[] args) { 
    System.out.println("Richards benchmark (deutsch_no_acc) starting...");
    long startTime = System.currentTimeMillis();
    if (!run())
      return;
    long endTime = System.currentTimeMillis();
    System.out.println("finished.");
    total_ms= endTime - startTime;
    System.out.println("Total time for " + iterations + " iterations: "
		       + (total_ms) + " ms");
    System.out.println("Average time per iteration: "
		       + (total_ms / iterations) + " ms");
  }

  static void schedule() {
    Task t = Task.taskList;
    while (t != null) {
      Packet pkt = null;

      if (Task.tracing) 
	System.out.println("tcb=" + t.id);

      if (t.IsTaskHoldingOrWaiting()) 
        t = t.link;
      else {
        if (Task.tracing) Task.trace((char)('0' + t.id));
        t = t.RunTask();
      }
    }
  }

  public boolean run() {
    for (int i= 0; i < iterations; i++){
      Task.holdCount= Task.qpktCount= 0;  // Added to allow repeated execution
					  // of the test.    Ole Agesen, 3/95.

      new IdleTask(I_IDLE, 1, 10000, (new TaskState()).Running(),
		   new IdleTaskRec());

      Packet wkq = new Packet(null, 0, K_WORK);
      wkq = new Packet(wkq, 0, K_WORK);
      new WorkTask(I_WORK, 1000, wkq,
		   (new TaskState()).WaitingWithPacket(),
		   new WorkerTaskRec());

      wkq = new Packet(null, I_DEVA, K_DEV);
      wkq = new Packet(wkq, I_DEVA, K_DEV);
      wkq = new Packet(wkq, I_DEVA, K_DEV);
      new HandlerTask(I_HANDLERA, 2000, wkq,
		      (new TaskState()).WaitingWithPacket(),
		      new HandlerTaskRec());

      wkq = new Packet(null, I_DEVB, K_DEV);
      wkq = new Packet(wkq, I_DEVB, K_DEV);
      wkq = new Packet(wkq, I_DEVB, K_DEV);
      new HandlerTask(I_HANDLERB, 3000, wkq,
		      (new TaskState()).WaitingWithPacket(),
		      new HandlerTaskRec());

      wkq = null;
      new DeviceTask(I_DEVA, 4000, wkq, (new TaskState()).Waiting(),
		     new DeviceTaskRec());
      new DeviceTask(I_DEVB, 5000, wkq, (new TaskState()).Waiting(),
		     new DeviceTaskRec());

      schedule();

      if (Task.qpktCount == 23246 && Task.holdCount == 9297) 
        ; // correct
      else {
        System.out.println("Incorrect results!");
        return false;
      }
    }
    return true;
  }

  // Task IDs
  static final int
    I_IDLE = 1,
    I_WORK = 2,
    I_HANDLERA = 3,
    I_HANDLERB = 4,
    I_DEVA = 5,
    I_DEVB = 6;

  // Packet types
  static final int
    K_DEV = 1000,
    K_WORK = 1001;
}

